---
title: "Coral Reefs - SNP genotyping"
author: "Chiara Duijser, Sage Rassmussen"
output: html_document
date: "2024-02-22"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Install packages & load library
```{r}
#install.packages("dartR")
#install.packages("dplyr")

#error when trying to import library for dartR. Need to install SNPRelate below
#if (!require("BiocManager", quietly = TRUE))
 # install.packages("BiocManager")

#BiocManager::install("SNPRelate")

# load library
library(dplyr)
library(dartR)
#gl.install.vanilla.dartR() # see github instillation info for more details on this

#install.packages("ggplot2")
library(ggplot2)
```


### Set working directory
```{r}
setwd("../SNP genotyping/Report-DPoc22-7677")
```


### Import data
```{r}
# create metafile.csv with first column as id = sample id's and pop = population/species names and then any additional data you want to include
gl <- gl.read.dart(filename="Report_DPoc22-7677_SNP_2.csv", ind.metafile="230317_ind_metrics_latlon.csv")

# to check genlight object is compliant run below code and see results to check for errors
gl.compliance.check(gl)
save(gl, file="gl_dartR.Rdata")
```


### Interrogate genlight object
```{r}
# some useful commands to get data from your genlight object
nLoc(gl) #number of loci = 36,706
locNames(gl) #list of loci
nInd(gl) #number of individuals (specimens / samples)  = 91
indNames(gl) #list of individuals
nPop(gl) #number of populations = 9 
popNames(gl) #list of populations
pop(gl) #list of population assignments for each individual
as.matrix(gl) #generate a matrix of the SNP scores, with 0 as homozygous reference, 2 as homozygous alternate, and 1 as heterozygous.
glPlot(gl) #a smear plot of individual against locus, useful for gross pattern identification and assessment of allelic dropout
gl <- readRDS("gl_dartR.Rdata")
```

## Filtering data
### Filter data step 1: checking original data based on Symbiodiniaceae genome 
```{r}
gl.blast(gl, "../symC_scaffold_40.fasta", verbose = 5) #NULL
gl.blast(gl, "../102_symbd_genome_scaffold.fa", verbose = 5) #NULL
gl.blast(gl, "../Breviolum_minutum.v1.0.genome.fa", verbose = 5)
   #   qseqid                      sacc                    stitle                                                            qseq
   # 2  31054 scaffold1595.1|size109920 scaffold1595.1|size109920 GACTAACTGACTAGCTAACTGACTAACTTATCGAC---CTACTGACTGACTGATTGTTTGACT
   # sseq nident mismatch pident length   evalue bitscore qstart qend sstart  send
   # 2 GACTAACTGACTAGCTAACTGACTAACTGACCAACTTGCTAATGACTAACTGAATGACTGACT     52        8  82.54     63 3.35e-06     52.8      8   67  45188 45250
   # gapopen gaps qlen   slen PercentageOverlap
   # 2       1    3   69 110003         0.9130435

gl.blast(gl, "../Cladocopium_goreaui_Genome.Scaffolds.fasta", verbose = 5) #NULL
gl.blast(gl, "../Fugacium_kawagutii_V1_genome_scaffold.fasta", verbose = 5) #NULL
gl.blast(gl, "../Fugacium_kawagutii_V2_genome_Scaffolds.fasta", verbose = 5) #NULL
gl.blast(gl, "../Fugacium_kawagutii_V3_genome_scaffold.fasta", verbose = 5) #NULL
gl.blast(gl, "../Genome.Scaffolds.fasta", verbose = 5) #NULL
gl.blast(gl, "../symA3_scaffold_37.fasta", verbose = 5) #NULL
gl.blast(gl, "../symbB.v1.0.genome.fa", verbose = 5)
     #   qseqid                      sacc                    stitle                                                            qseq
     # 2  31054 scaffold1595.1|size109920 scaffold1595.1|size109920 GACTAACTGACTAGCTAACTGACTAACTTATCGAC---CTACTGACTGACTGATTGTTTGACT
     # sseq nident mismatch pident length   evalue bitscore qstart qend sstart  send
     # 2 GACTAACTGACTAGCTAACTGACTAACTGACCAACTTGCTAATGACTAACTGAATGACTGACT     52        8  82.54     63 3.35e-06     52.8      8   67  45188 45250
     # gapopen gaps qlen   slen PercentageOverlap
     # 2       1    3   69 110003         0.9130435

gl.blast(gl, "../SymbC1.MITO_seqs.fa", verbose = 5) #NULL
gl.blast(gl, "../Symbiodinium_microadriacticum_genome.scaffold.fasta", verbose = 5) #NULL 

#checking filtered data against on Symbiodiniaceae genome 
gl.blast(gl5, "../Breviolum_minutum.v1.0.genome.fa", verbose = 5)
gl.blast(gl5, "../symbB.v1.0.genome.fa", verbose = 5)

#NOTE: NO SYMBIONT CONTAMINATION ON FILTERED DATA.
```


### Remove technical replicates
```{r}
# First remove technical replicates by removing individuals with most missing data from each replicate set
gl.report.callrate(gl, method="ind") #reports missing data by individual

# Drop individuals with most missing data based on id=ind_names column in Excel!
ind_drop_list_2 <- c("18", "19.1", "20", "21", "22.1", "23", "24", "25.1", "26", "26.1",
                   "28", "29", "30", "30.1", "31.1", "32.1", "34", "36", "36.1", 
                   "37.1", "38.1", "39", "40.1", "41", "42.1", "43", "44", "45.1", "46", "46.2",
                   "48", "48.1", "50.1", "51", "52", "53", "53.1")

gl4 <- gl.drop.ind(gl, ind.list=ind_drop_list_2)
gl4

gl.smearplot(gl4)
```


### Remove additional samples on the run that are not needed for my analysis
```{r}
gl5 <- gl.drop.pop(gl4, pop.list="Supervisor")
gl5 <- gl.drop.pop(gl5, pop.list="Extra")
gl5 <- gl.drop.pop(gl5, pop.list="Opal - acuta")
gl5 <- gl.drop.pop(gl5, pop.list="Opal - verrucosa")
gl5 <- gl.drop.pop(gl5, pop.list="Opal - damicornis")

gl5 <- gl.recalc.metrics(gl5)
gl5

gl5 <- gl.filter.monomorphs(gl5)

popNames(gl5)
indNames(gl5)

saveRDS(gl5, file="gl5_dartR.Rdata")
# read it in again with
gl5 <- readRDS("gl5_dartR.Rdata")
```


### Filter data step 2
```{r}
gl5 <- gl.filter.secondaries(gl5)

gl5 <- gl.filter.reproducibility(gl5, t=0.995, verbose = 3)

gl5 <- gl.filter.callrate(gl5, method="loc", threshold=0.95, verbose = 3)

gl5 <- gl.filter.rdepth(gl5, lower=5, verbose = 3)

gl5 <- gl.filter.maf(gl5, threshold=0.05, verbose = 3)

gl5 <- gl.filter.monomorphs(gl5, verbose = 3)

gl5 <- gl.recalc.metrics(gl5, verbose = 3)

gl5 <- gl.impute(gl5, method="neighbour", verbose = 3)

gl5 <- gl.recalc.metrics(gl5)

gl5

saveRDS(gl5, file="gl5_dartR.Rdata")
# read it in again with
gl5 <- readRDS("gl5_dartR.Rdata")
indNames(gl5)
```


## Analyse data
### Test for clones
```{r}
gl5_grm <- gl.grm(gl5, plotheatmap = FALSE,
                  verbose = 3)

gl.grm.network(gl5_grm, gl5, method = "mds", link.size = 2, relatedness_factor = 0.125)

gl.grm.network(
  gl5_grm,
  gl5,
  method = "fr",
  node.size = 8,
  node.label = TRUE,
  node.label.size = 2,
  node.label.color = "black",
  link.color = NULL,
  link.size = 2,
  relatedness_factor = 0.125,
  title = "Network based on a genomic relationship matrix",
  palette_discrete = NULL,
  save2tmp = FALSE,
  verbose = NULL
)
```


### Remove clones
```{r}
# Genomic relationship matrix identified 20.1 and 21.1 as clones. Will remove 20.1 as it had more missing data (negligible)
ind_drop_list3 <- c("20.1")

gl5 <- gl.drop.ind(gl5, ind.list=ind_drop_list3)
gl5

# After removing, resultant dataset may contain monomorphic loci, so recalculate locus metrics again
gl5 <- gl.filter.monomorphs(gl5, verbose = 3)
gl5 <- gl.recalc.metrics(gl5, verbose = 3)

saveRDS(gl5, file="gl5_CLONE_removed.Rdata")

# Genlight object that STRUCTURE was performed on
gl5 <- readRDS("../SNP genotyping/Report-DPoc22-7677/gl5_CLONE_removed.Rdata")
gl5
indNames(gl5)
```



### Assign new populations
Re-ran population genetic analysis (PCA, Fst, AMOVA) on four population models:

*Model_1 = populations are reef locations
*Model_2 = populations are DAPC lineages
*Model_3 = populations are STR lineages including ‘hybrid’ population
*Model_4 = populations are STR lineages with hybrids removed

 
```{r}
### Model 1
popNames(gl5) 
indNames(gl5)

### Model 2
# new population *glp2* based on mangrove and reef str/dapc populations
glp2 <- gl.define.pop(gl5, ind.list=c("19", "21.1", "22", "23.1", "24.1", "25", "26.2", "27", "27.1", "28.1", "29.1", "31", "32", "33", "35", "35.1", "36.2", "37", "38", "39.1", "40", "41.1")
                      , new='Population 2') # Mangrove
glp2 <- gl.define.pop(glp2, ind.list=c("31", "32", "33", "35", "35.1", "36.2", "37", "38", "39.1", "40", "41.1")
                      , new='Population 1') # Reef
popNames(glp2)

### Model 3
# new population *glp3* based on mangrove and reef str/dapc populations with STR hybrids defined
glp3 <- gl.define.pop(glp2, ind.list=c("21.1", "35.1", "36.2", "37", "38", "40")
                     , new='Hybrid')
popNames(glp3)
glp3
saveRDS(glp3, file="glp3_hybrids_included.Rdata")

gl.tree.nj(glp3, type="phylogram", labelsize=1.0)
gl.tree.nj(glp3, type="cladogram", labelsize=1.0) # cladogram, fan, unrooted

### Model 4
# new population *glp4* based on mangrove and reef str/dapc populations with STR hybrids removed
glp4 <- gl.drop.pop(glp3, pop.list="Hybrid")
### With the dropping of Hybrids, resultant dataset contained monomorphic loci
### I filter on monomorphs and recalculate locus metrics
glp4 <- gl.filter.monomorphs(glp4, verbose = 3) 
glp4 <- gl.recalc.metrics(glp4)
glp4
popNames(glp4)
```


### PCA: Run analysis on new populations (STR/DAPC) and old populations (reef)
#### Model 3
```{r}
pcoa_gl5_model3 <- gl.pcoa(glp3)

col <- c("orange", "#1F78B4", "#33A02C")
#shp <- gl.select.shapes(select=c(16,17,18))

gl.pcoa.plot(pcoa_gl5_model3, glp3, pt.colors=col, pt.size=3, pop.labels="pop", save2tmp = TRUE)

#gl.pcoa.plot(pcoa,test,save2tmp = TRUE)
# listing the plots and tables in tempdir
gl.list.reports()
# choosing the plot or table to retrieve
pca_model3 <- gl.print.reports(16)
# modifying the plot
(pca_model3 <- pca_model3 + theme_bw() + 
    labs(x=substitute(paste(bold('PC1 (61.4%)'))), y = substitute(paste(bold('PC2 (14.7%)'))), fill="Population"))
#ggsave("PCA_host_genetics_model3_pop.pdf", width=6, height=4)
```

#### Model 4 (Supplementary)
```{r}
pcoa_gl5_model4 <- gl.pcoa(glp4)

#install.packages("RColorBrewer")
library(RColorBrewer)
#brewer.pal(6,"Paired")
#display.brewer.pal(4,"Paired")

col <- gl.select.colors(library='brewer',palette='Paired',ncolors=4, select=c(4,2,3,1))
shp <- gl.select.shapes(select=c(16,17,16,17))
gl.pcoa.plot(pcoa_gl5_model4, glp4, pt.colors=col, pt.shapes=shp, pt.size=3, pop.labels="pop", save2tmp = TRUE) 
#gl.pcoa.plot(pcoa,test,save2tmp = TRUE)
# listing the plots and tables in tempdir
gl.list.reports()
# choosing the plot or table to retrieve
pca_model4 <- gl.print.reports(23)
# modifying the plot
(pca_model4 <- pca_model4 + theme_bw() + 
    labs(x=substitute(paste(bold('PC1 (97.8%)'))), y = substitute(paste(bold('PC2 (0.3%)'))), fill="Population"))
#ggsave("PCA_host_genetics_model4.pdf", width=6, height=4)
```


### Genetic distance / Fst
```{r}
#(Fst <- gl.fst.pop(glp2))
#Fst$Pvalues

# run Fst on original sites (Model 1)
(Fst_p1 <- gl.fst.pop(gl5, nboots = 999))

# run Fst between 2x populations (Model 2)
(Fst_p2 <- gl.fst.pop(glp2, nboots = 999))

# run Fst between 3x populations (Model 3)
(Fst_p3 <- gl.fst.pop(glp3, nboots = 999))

# run Fst between 2x populations (hybrids removed) (Model 4)
(Fst_p4 <- gl.fst.pop(glp4, nboots = 999))
```


### DAPC
Check to see how many populations assumed by DAPC

```{r}
dapc <- find.clusters(gl5)
dapc # K=2

# $Kstat
# K=1      K=2 
# 126.9056 111.6733 

# $stat
# K=2 
# 111.6733 

# $grp
# 19 21.1   22 23.1 24.1   25   26   27   28 29.1 30.1   31   32   33   34   35 36.2   37   38 39.1   40 41.1 
# 1    2    1    1    1    1    1    1    1    1    1    2    2    2    2    2    2    2    2    2    2    2 
# Levels: 1 2

# $size
# [1] 10 12
```


### AMOVA
```{r}
library(poppr)
### Model 3
# Converting genlight to genind 
geni_model3 <- gl2gi(glp3)
# Assign pop information to strata
strata(geni_model3) <- data.frame(pop = pop(geni_model3))
# Performing AMOVA
amova_poppr_model3 <- poppr.amova(geni_model3, ~pop, nperm = 1000)
amova_poppr_model3

# Test for significance 
amova_sig_m3 <- randtest(amova_poppr_model3)
plot(amova_sig_m3)
amova_sig_m3

### Model 4
# Converting genlight to genind 
geni_model4 <- gl2gi(glp4)
# Assign pop information to strata
strata(geni_model4) <- data.frame(pop = pop(geni_model4))
# Performing AMOVA
amova_poppr_model4 <- poppr.amova(geni_model4, ~pop, nperm = 1000)
amova_poppr_model4

# Test for significance 
amova_sig_m4 <- randtest(amova_poppr_model4)
plot(amova_sig_m4)
amova_sig_m4
```


### Convert to STRUCTURE format (if performing STRUCTURE outside of R)
```{r}
gl2structure(gl5, outpath = "..")

#nLoc(gl5) #number of loci
# How many loci: 2148
#nInd(gl5) #number of individuals (specimens or samples) 
# How many ind: 22
```


### Install packages for importing and reading STRUCTURE output
```{r}
#install devtools package from CRAN
#install.packages('devtools',dependencies=T)
library(devtools)

#install pophelper package from GitHub
#install_github('royfrancis/pophelper')

#load library for use
library(pophelper)

# check version
packageDescription("pophelper", fields="Version")
```


### Setting working directory to correct STRUCTURE output folder
For your STRUCTURE results, make sure all runs for all K are put in on directory. Mine are all in 'results_all'.

```{r}
setwd("../SNP genotyping/STRUCTURE_pop/results_all")

# import data from directory
sfiles <- c("gl_k1_run1_f", "gl_k1_run2_f", "gl_k1_run3_f", "gl_k1_run4_f", "gl_k1_run5_f", ##K=1, 5 runs, file 1-5
            "gl_k2_run1_f", "gl_k2_run2_f", "gl_k2_run3_f", "gl_k2_run4_f", "gl_k2_run5_f", ##K=2, 5 runs, file 6-10
            "gl_k3_run1_f", "gl_k3_run2_f", "gl_k3_run3_f", "gl_k3_run4_f", "gl_k3_run5_f", ##K=3, 5 runs, file 11-15
            "gl_k4_run1_f", "gl_k4_run2_f", "gl_k4_run3_f", "gl_k4_run4_f", "gl_k4_run5_f", ##K=4, 5 runs, file 16-20
            "gl_k5_run1_f", "gl_k5_run2_f", "gl_k5_run3_f", "gl_k5_run4_f", "gl_k5_run5_f") ##K=5, 5 runs, file 21-25

slist <- readQ(files=sfiles,filetype="structure")

# a few functions to check data below. 
# check class of ouput
class(slist)
# view head of first converted file
head(slist[[1]])

# qlist attributes
attributes(slist)
# dataframe attributes
attributes(slist[[1]])
# attributes of all runs
# lapply(slist,attributes)
```


### Evanno plot
```{r}
#evannoMethodStructure(data=sr1,exportplot=T,exportpath=getwd()) # sr1 not found!!!

#Plot saves to your working directory, open to visualise and choose optimal K.
```


### Plot
```{r}
# plot selected runs (Change the numbers to choose the K you want to plot as below)
p1 <- plotQ(slist[c(6:10)],imgoutput="join",returnplot=T,exportplot=F,basesize=11)
p1

# aligk and plot selected runs (as above)
slist1 <- alignK(slist[c(6:10)])
p1 <- plotQ(slist1,imgoutput="join",returnplot=T,exportplot=F,basesize=11)
p1

#merge aligned runs and plot 
mlist <- mergeQ(slist1)
p2 <- plotQ(qlist = mlist,returnplot = T, exportplot = F)
p2

# customised plot for merged run below: 

(p3 <- plotQ(qlist = mlist, 
            sortind="all", # test options "all", "Clussp71", "Clussp72"...
            clustercol=c("#84b4f0ff", "#ccdc7a80", "#97020080", "#068f9b80", "#cc3fd880"), #add more colours if you have more populations. Colour choice based on own preferences to plot. 
            sharedindlab = FALSE,
            showtitle = TRUE, 
            titlelab = "ALL K=2", #update depending on your K
            showindlab = TRUE,
            basesize = 11,
            showyaxis = TRUE,
            returnplot = TRUE, 
            exportplot = FALSE))

# set colours
# ALL: Outlier, sp7cinthus, sp7sus, pectinata, sp7: 
#clustercol=c("#84b4f0ff", "#97020080", "#cc3fd880", "#068f9b80", "#ccdc7a80"),
## PEC: P1, P2
#clustercol=c("#068f9b80", "#068f1c99"),
# sp7: P1, P2
#clustercol=c("#97020080", "#96598590"),
# sp7:
#clustercol=c("#cc3fd880", "#cc8fc990", "#92011450",)
# sp7:
#clustercol=c("#ccdc7a80", "#B4F02B"),
# Export and save as svg file when happy with format. 
```


I choose to put results "mlist" in .csv file so I can easily adjust the figure in R

### STRUCTURE plot
```{r}
str_results <- read.csv("../SNP genotyping/STRUCTURE_pop/structure_results2.csv", stringsAsFactors=TRUE)

str(str_results)
str_results$name_DArT_wrong <- as.factor(str_results$name_DArT_wrong)
str_results$name_DArT_wrong <- factor(str_results$name_DArT_wrong, levels=c(
    "39.1", "41.1", "36.2", "40", "38", "37",  "31", "32", "33", "35", "35.1", 
    "25", "26.2", "27", "27.1", "28.1", "29.1", "19", "22", "23.1", "24.1", "21.1"
))

levels(str_results$name_DArT_wrong)
str_results$site <- factor(str_results$site, levels=c("Outer reef", "Inner reef", "Outer mangrove", "Inner mangrove"))

# Figure STRUCTURE adjusted in R so no need to export and edit in Adobe Illustrator
ggplot(str_results, aes(fill=cluster, y=value, x=name_DArT_wrong)) +
  geom_bar(position="fill", stat="identity") +
  facet_wrap(~ site, nrow = 1, scales = "free_x") +
  scale_fill_manual(values=c("#1F78B4", "#33A02C")) +
  labs(
    y="Admixture proportion", 
    x="",
    fill="Cluster") +
  theme_bw() +
  theme(legend.position = "bottom",
        axis.text.x=element_blank(), 
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(size=10, face="bold"),
        panel.border = element_blank(), 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), 
        axis.ticks = element_blank(),
        strip.text.x = element_text(size = 11, face="bold")
        )
#ggsave("StructurePlot_K2_NEW.pdf", width = 9, height = 3) #nrow
```

```{r}
str_results_pie <- read.csv("../SNP genotyping/STRUCTURE_pop/structure_results_pie.csv", stringsAsFactors=TRUE)
str(str_results_pie)
str_results_pie$Cluster <- as.factor(str_results_pie$Cluster)
str_results_pie$pop <- as.factor(str_results_pie$pop)
str_results_pie$pop <- factor(str_results_pie$pop, levels=c("Outer reef", "Inner reef", "Outer mangrove", "Inner mangrove"))

# Figure 2B
ggplot(str_results_pie, aes(x = "", y = Value, fill = Cluster)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar(theta = "y") +
  facet_wrap(~ pop, ncol = 2) +
  scale_fill_manual(values = c("#33A02C", "#1F78B4")) +
  theme_void()
#ggsave("Structure_Pie_NEW.pdf", width=6, height=6)
```










